File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        //keep element in the first position
        //move the index to find all combinations with num in that place
        //when idx reaches the end, add nums to ans
        //swap them back to keep nums in its original place
        
        helper(0,ans,nums); //to handle the backtracking
        
        return ans;
    }
    
    void helper(int idx,vector<vector<int>>& ans,vector<int>& nums){
        if(idx == nums.size()){
            ans.push_back(nums);
        }
        
        for(int i=idx;i<nums.size();i++){
            swap(nums[idx],nums[i]);
            helper(idx+1,ans,nums);
            swap(nums[idx],nums[i]);
        }
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        # make copies
        def dfs(arr,nums):
            if len(nums) == 0:
                ans.append(arr.copy())
                return
            
            for i in range(len(nums)):
                dfs(arr + [nums[i]], nums[:i] + nums[i+1:])
        
        dfs([],nums)

        return ans